File: src/main/java/com/br/service/S3Service.java
package com.br.service;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import software.amazon.awssdk.core.sync.RequestBody;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.*;

import java.util.Map;

@ApplicationScoped
public class S3Service {

    @Inject
    S3Client s3;

    @ConfigProperty(name = "app.raw-bucket")
    String rawBucket;

    @ConfigProperty(name = "app.thumbnails-bucket")
    String thumbnailsBucket;

    /**
     * Returns the user-defined metadata of an object in the raw bucket.
     * S3 strips the "x-amz-meta-" prefix; keys are lowercased.
     * e.g. x-amz-meta-target-size → "target-size"
     */
    public Map<String, String> getObjectMetadata(String key) {
        HeadObjectResponse response = s3.headObject(
            HeadObjectRequest.builder()
                .bucket(rawBucket)
                .key(key)
                .build()
        );
        return response.metadata();
    }

    /**
     * Returns true if the output key already exists in the thumbnails bucket.
     * Used as the idempotency gate before processing.
     */
    public boolean thumbnailExists(String outputKey) {
        try {
            s3.headObject(HeadObjectRequest.builder()
                .bucket(thumbnailsBucket)
                .key(outputKey)
                .build());
            return true;
        } catch (NoSuchKeyException e) {
            return false;
        }
    }

    public byte[] downloadRaw(String key) {
        return s3.getObjectAsBytes(
            GetObjectRequest.builder()
                .bucket(rawBucket)
                .key(key)
                .build()
        ).asByteArray();
    }

    public void uploadThumbnail(String key, byte[] data, String contentType) {
        s3.putObject(
            PutObjectRequest.builder()
                .bucket(thumbnailsBucket)
                .key(key)
                .contentType(contentType)
                .build(),
            RequestBody.fromBytes(data)
        );
    }
}

File: app/services/s3_service.py
import os
import boto3
from botocore.exceptions import ClientError


class S3Service:
    def __init__(self) -> None:
        self._raw_bucket = os.environ["RAW_BUCKET"]
        self._thumbnails_bucket = os.environ["THUMBNAILS_BUCKET"]
        self._client = boto3.client("s3")

    def get_object_metadata(self, key: str) -> dict[str, str]:
        """Returns user-defined metadata for an object in the raw bucket.
        S3 strips the 'x-amz-meta-' prefix and lowercases all keys.
        """
        response = self._client.head_object(Bucket=self._raw_bucket, Key=key)
        return response.get("Metadata", {})

    def thumbnail_exists(self, output_key: str) -> bool:
        """Idempotency gate: returns True if the output key already exists."""
        try:
            self._client.head_object(Bucket=self._thumbnails_bucket, Key=output_key)
            return True
        except ClientError as e:
            if e.response["Error"]["Code"] == "404":
                return False
            raise

    def download_raw(self, key: str) -> bytes:
        response = self._client.get_object(Bucket=self._raw_bucket, Key=key)
        return response["Body"].read()

    def upload_thumbnail(self, key: str, data: bytes, content_type: str) -> None:
        self._client.put_object(
            Bucket=self._thumbnails_bucket,
            Key=key,
            Body=data,
            ContentType=content_type,
        )
